README.md
The lightweight Flash video player for Video.js. This allows the Video.js player's skins, plugins, and other features to work with both HTML5 and Flash.

This project doesn't need to be used if you simply want to use the Flash tech in Video.js.

- For Video.js 5.x and below, the Flash tech is part of the [Video.js core repository](https://github.com/videojs/video.js).
- For Video.js 6.x and above, the Flash tech is in a [separate repository](https://github.com/videojs/videojs-flash).

## Installation

[![Greenkeeper badge](https://badges.greenkeeper.io/videojs/video-js-swf.svg)](https://greenkeeper.io/)

1. Install Node Packages.
```bash
    npm install
   ```
2. Compile SWF.
Development (places new SWF in /dist/):
```bash
    grunt mxmlc
   ```
Production/ Distribution (runs mxmlc task and copies SWF to dist/):
```bash
    grunt dist
   ```
3. Run Connect Server.
```bash
    grunt connect:dev
```
4. Open your browser at [http://localhost:8000/index.html](http://localhost:8000/index.html) to see a video play.  You can keep using grunt to rebuild the Flash code.

## Releasing

1. Make sure that the following file is modified with these values:

```
node_modules/flex-sdk/lib/flex_sdk/frameworks/flex-config.xml
```

```xml
<!-- Specifies the minimum player version that will run the compiled SWF. -->
<target-player>10.3</target-player>

<!-- Specifies the version of the compiled SWF -->
<swf-version>12</swf-version>
```

2. Run the commands:
```sh
npm version {major,minor,patch}
npm publish
```
The swf and changelog will be automatically built and added to the repo on version.

## Running Unit and Integration Tests

** Note - We want to drop all of this for grunt based / Karma testing.

For unit tests, this project uses [FlexUnit](http://flexunit.org/). The unit tests can be found in [project root]/src/com/videojs/test/

For integration tests, this project uses [qunit](http://qunitjs.com/). The integration tests can be found in [project root]/test

In order to run all of the tests, use the links at  [http://localhost:8000/index.html](http://localhost:8000/index.html)

There are very few tests.  Adding to them is a fantastic and much appreciated way to contribute.


CHANGELOG.md
CHANGELOG
=========

## HEAD (Unreleased)
* Add license to package.json

--------------------

## 5.4.2 (2018-09-20)
* Fix video dimensions pinned to 100x100

## 5.4.1 (2017-07-25)
* chore: update broken URLs
* fix: decouple muted() from volume=0

## 5.4.0 (2017-04-28)
* Add getVideoPlaybackQuality API
* Fix buffered() getter for RTMP provider
* Fix paused state for RTMP provider after video ends

## 5.3.0 (2017-02-17)
* @albertogasparin added FCsubscribe call and proxy type to RTMP video provider
* @mjneil add callback to adjust currentTime for video provider

## 5.2.0 (2017-02-07)
* @mjneil added appendChunkReady, another way of passing data into the swf

## 5.1.0 (2016-07-18)
* @alex-philips added support for the onTextData event

## 5.0.3 (2016-05-31)
* Fix `muted` and `loop` attributes not being passed to the swf ([view](https://github.com/videojs/video-js-swf/pull/205))

## 5.0.2 (2016-05-06)
* If we are in data generation mode and even if `_playbackStarted` hasn't happened yet, we should still set `_isSeeking` to true so that we can correctly emit `seeked` after an initial (before playback) seek ([view](https://github.com/videojs/video-js-swf/pull/204))

## 5.0.1 (2015-11-06)
* Fix an issue where the player would not report is was seeking after ending ([view](https://github.com/videojs/video-js-swf/pull/192))

## 5.0.0 (2015-10-28)
* Return an empty set of seekable time ranges when seeking in data generation mode ([view](https://github.com/videojs/video-js-swf/pull/187))

## 5.0.0-rc1 (2015-07-27)
* Don't fire loadstart or loadedmetadata in data generation mode ([view](https://github.com/videojs/video-js-swf/pull/178))
* Remove unused poster support ([view](https://github.com/videojs/video-js-swf/pull/182))
* Expose buffered as ranges, not just an end point ([view](https://github.com/videojs/video-js-swf/pull/180))
* Do not seek to to the beginning after a video ends ([view](https://github.com/videojs/video-js-swf/pull/172))

## 5.0.0-rc0 (2015-07-21)
* Let javascript fire "seeking" instead of handling it in the SWF ([view](https://github.com/videojs/video-js-swf/pull/171))

## 4.7.2 (2015-06-30)
* Fixed an issue where an image from previous video could flash briefly while a new video is loaded ([view](https://github.com/videojs/video-js-swf/pull/167))

## 4.7.1 (2015-06-23)
* Fixed an issue where playback required two clisk to start when preload was not auto
* @qpSHiNqp fix issue that would cause incorrect aspect ratios for some videos ([view](https://github.com/videojs/video-js-swf/pull/165))

## 4.7.0 (2015-05-19)
* @bc-bbay the preload attribute should be a string, not a boolean ([view](https://github.com/videojs/video-js-swf/pull/160))
* @Wellming fix manual tests ([view](https://github.com/videojs/video-js-swf/pull/154))

## 4.6.1 (2015-04-22)
* @bclwhitaker append END_SEQUENCE properly in data generation mode ([view](https://github.com/videojs/video-js-swf/pull/152))

## 4.6.0 (2015-04-16)
* Add vjs_discontinuity ([view](https://github.com/videojs/video-js-swf/pull/150))
* Don't call resume() on NetstreamPlayStart ([view](https://github.com/videojs/video-js-swf/pull/147))

## 4.5.4 (2015-03-17)
* Improved handling of the paused state, and the loadstart and canplay events ([view](https://github.com/videojs/video-js-swf/pull/139))
* Fixed a potential XSS issue with the swf event callbacks ([view](https://github.com/videojs/video-js-swf/pull/143))
* Prevented pause from firing after eneded ([view](https://github.com/videojs/video-js-swf/pull/144))

## 4.5.3 (2015-01-22)
* Paused should be true before a source has been set by default

## 4.5.2 (2014-12-04)
* Fixed an issue where Flash would crash when switching sources quickly ([view](https://github.com/videojs/video-js-swf/pull/131))

## 4.5.1 (2014-10-15)
* Fixed an issue where changing the source immediately after seeking could cause an error ([view](https://github.com/videojs/video-js-swf/pull/125))
* Added sanitation for all data that might be passed through the external interface ([view](https://github.com/videojs/video-js-swf/pull/127))

## 4.5.0 (2014-09-29)
* Buffering and playback event fixes ([view](https://github.com/videojs/video-js-swf/pull/122))

## 4.4.5 (2014-09-25)
* Fixed sanitation of URLs to special case blob URLs ([view](https://github.com/videojs/video-js-swf/pull/121))

## 4.4.4 (2014-09-22)
* Added sanitizing of the src param ([view](https://github.com/videojs/video-js-swf/pull/120))

## 4.4.3 (2014-08-14)
* Rebuild with Flash target-player 10.3 and swf-version 12. ([view](https://github.com/videojs/video-js-swf/issues/113))

## 4.4.2 (2014-07-11)
* Fixed networkState reporting to be more accurate after loadstart ([view](https://github.com/videojs/video-js-swf/pull/106))

## 4.4.1 (2014-06-11)
* Ignore unnecessary files from npm packaging ([view](https://github.com/videojs/video-js-swf/pull/87))
* Fixed bug triggering `playing` ([view](https://github.com/videojs/video-js-swf/pull/90))
* Fixed bug with the timing of `loadstart` ([view](https://github.com/videojs/video-js-swf/pull/93))
* Added support for clearing the NetStream while in Data Generation Mode ([view](https://github.com/videojs/video-js-swf/pull/93))
* Fixed silent exception when opening MediaSources ([view](https://github.com/videojs/video-js-swf/pull/97))

## 4.4.0 (2014-02-18)
* Added changelog
* Added support for using NetStream in Data Generation Mode ([view](https://github.com/videojs/video-js-swf/pull/80))
* Extended base support for external appendData for integration with HLS / Media Source plugins ([view](https://github.com/videojs/video-js-swf/pull/80))
* Fixed bug with viewport sizing on videos which don't present meta data ([view](https://github.com/videojs/video-js-swf/pull/80))
* Fixed bugs with buffered and duration reporting on non-linear streams ([view](https://github.com/videojs/video-js-swf/pull/80))
* Added refined seeking for use on non-linear streams ([view](https://github.com/videojs/video-js-swf/pull/80))

test.sh
#!/bin/bash

# See README.md for instructions

pwd=`pwd`
flex_sdk=/Applications/Flex

# use command-line parameters if the default location
# isn't available
if [ ! -d "$flex_sdk" ]
then
  flex_sdk=$1
fi

# if we needed a command-line parameter and didn't get one,
# complain about it
if [ -z "$flex_sdk" ]
then
  echo "Usage: test.sh flex_sdk_dir"
  exit 1
fi

# complain about the incorrect directory passed in
if [ ! -d "$flex_sdk" ]
then
  echo "Flex SDK not found at $flex_sdk"
  exit 1
fi

# and now we can get started, echoing out what we're doing

echo "Compiling bin-debug/test-video-js.swf..."
$flex_sdk/bin/mxmlc ./src/AirTestRunner.mxml -o ./bin-debug/test-video-js.swf -use-network=false -static-link-runtime-shared-libraries=true -library-path+=libs

forever stop http/server.js 2> /dev/null
forever start --minUptime=1000 --spinSleepTime=1000 http/server.js

echo "Running test-video-js.swf..."
open bin-debug/test-video-js.swf
echo "Open command was run.  If you don't see a SWF running, you're"
echo "probably not on a Mac.  Try to run this:"
echo "$pwd/bin-debug/test-video-js.swf"

echo "Running tests/test.html..."
open http://localhost:8000/tests/test.html
echo "Again, if you don't see the page above running, you'll need"
echo "to open it manually in a browser:"
echo "http://localhost:8000/tests/test.html"


tests/test.js
var uid = 0;

function createSWF(e){

  var flashvars = {
    readyFunction: "onSWFReady",
    eventProxyFunction: "onSWFEvent",
    errorEventProxyFunction: "onSWFErrorEvent",
    src: "http://vjs.zencdn.net/v/oceans.mp4",
    autoplay: false,
    preload: 'auto',
    poster: "http://vjs.zencdn.net/v/oceans.png"
  };

  var params = {
    allowScriptAccess: "always",
    bgcolor: "#000000"
  };

  var attributes = {
    id: "videoPlayer"+uid,
    name: "videoPlayer"+uid
  };

  swfobject.embedSWF("../dist/video-js.swf", "videoPlayer"+uid, "100%", "100%", "10.3", "", flashvars, params, attributes);
}

function swfSetup(){
  stop();

  window.onSWFReady = $.proxy(function(swfID){
    // console.log("onSWFReady", swfID);
    this.box = document.getElementById(swfID+"_box");
    this.swf = document.getElementById(swfID);
    start();
  }, this);

  window.onSWFEvent = function(swfID, eventName){
    console.log("onSWFEvent", swfID, eventName);

    // Triggering on outer div because triggering in object wasn't working for some reason.
    $("#"+swfID+"_box").trigger(eventName);
  };

  // Bind events to outer box.
  this.on = function(eventName, fn){
    $(this.box).bind(eventName, $.proxy(fn, this));
  }

  // Custom methods for failing a test after a certain amount of time
  this.failIn = function(ms){ this.failOutID = setTimeout(function(){ start(); }, ms); };
  this.cancelFail = function(){ clearTimeout(this.failOutID); };

  // Embed new box and placeholder for swf
  uid++;
  var id = "videoPlayer"+uid;
  $("#custom-fixture").append("<div id='"+id+"_box' class='box'><div id='"+id+"'></div></div>")

  createSWF();
}

function swfTeardown(){
  swfobject.removeSWF(this.swf.id);
  delete this.swf;
  $("#custom-fixture").html("");
}


module("SWF Tests", {
  setup: swfSetup,
  teardown: swfTeardown
});

test("SWF Set Up & Ready", 2, function() {
  ok(this.swf, "Player has been set up and is ready.");
  ok(typeof this.swf.vjs_play == 'function', "API Methods are available on ready.");
});

// Pause after playing event
test("Pause After 'playing' Event", 1, function() {
  stop();

  // Wait for playing even then call pause()
  this.on("playing", function(){
    this.swf.vjs_pause();
  });

  this.on("pause", function(){
    ok(true, "Player pauses.");

    this.cancelFail();
    start();
  });

  // Fail after 5 seconds if it doesn't work
  this.failIn(3000);

  this.swf.vjs_play();
});


// Play Method
test("Play", 1, function() {
  stop();

  this.on("playing", function(){
    ok(true, "Player plays.");
    start();
  });

  this.swf.vjs_play();
});

// 'seeked' event
// Commented-out as this only works intermittently.  Sometimes the "seeked"
// event doesn't happen
/*test("Seeked event fires after time change", 1, function() {
  stop();
  this.on("loadeddata", function(){
    this.swf.vjs_pause();
    this.ct = this.swf.vjs_getProperty("currentTime");

    this.on("seeked", function(){

      ok(this.ct != this.swf.vjs_getProperty("currentTime"), "currentTime changed");
      start();
    });

    this.swf.vjs_setProperty("currentTime", 30);
  });

  this.swf.vjs_play();
});
*/

// Commented out for unknown reasons...

// /* Methods
// ================================================================================ */
// module("API Methods", {
//   setup: playerSetup,
//   teardown: playerTeardown
// });
//
// function failOnEnded() {
//   this.player.one("ended", _V_.proxy(this, function(){
//     start();
//   }));
// }
//
// // Play Method
// test("play()", 1, function() {
//   stop();
//
//   this.player.one("playing", _V_.proxy(this, function(){
//     ok(true);
//     start();
//   }));
//
//   this.player.play();
//
//   failOnEnded.call(this);
// });
//
// // Pause Method
// test("pause()", 1, function() {
//   stop();
//
//   // Flash doesn't currently like calling pause immediately after 'playing'.
//   this.player.one("timeupdate", _V_.proxy(this, function(){
//
//     this.player.pause();
//
//   }));
//
//   this.player.addEvent("pause", _V_.proxy(this, function(){
//     ok(true);
//     start();
//   }));
//
//   this.player.play();
// });
//
// // Paused Method
// test("paused()", 2, function() {
//   stop();
//
//   this.player.one("timeupdate", _V_.proxy(this, function(){
//     equal(this.player.paused(), false);
//     this.player.pause();
//   }));
//
//   this.player.addEvent("pause", _V_.proxy(this, function(){
//     equal(this.player.paused(), true);
//     start();
//   }));
//
//   this.player.play();
// });
//
// test("currentTime()", 1, function() {
//   stop();
//
//   // Try for 3 time updates, sometimes it updates at 0 seconds.
//   // var tries = 0;
//
//   // Can't rely on just time update because it's faked for Flash.
//   this.player.one("loadeddata", _V_.proxy(this, function(){
//
//     this.player.addEvent("timeupdate", _V_.proxy(this, function(){
//
//       if (this.player.currentTime() > 0) {
//         ok(true, "Time is greater than 0.");
//         start();
//       } else {
//         // tries++;
//       }
//
//       // if (tries >= 3) {
//       //   start();
//       // }
//     }));
//
//   }));
//
//   this.player.play();
// });
//
//
// test("currentTime(seconds)", 2, function() {
//   stop();
//
//   // var afterPlayback = _V_.proxy(this, function(){
//   //   this.player.currentTime(this.player.duration() / 2);
//   //
//   //   this.player.addEvent("timeupdate", _V_.proxy(this, function(){
//   //     ok(this.player.currentTime() > 0, "Time is greater than 0.");
//   //
//   //     this.player.pause();
//   //
//   //     this.player.addEvent("timeupdate", _V_.proxy(this, function(){
//   //       ok(this.player.currentTime() == 0, "Time is 0.");
//   //       start();
//   //     }));
//   //
//   //     this.player.currentTime(0);
//   //   }));
//   // });
//
//   // Wait for Source to be ready.
//   this.player.one("loadeddata", _V_.proxy(this, function(){
//
//     _V_.log("loadeddata", this.player);
//     this.player.currentTime(this.player.duration() - 1);
//
//   }));
//
//   this.player.one("seeked", _V_.proxy(this, function(){
//
//     _V_.log("seeked", this.player.currentTime())
//     ok(this.player.currentTime() > 1, "Time is greater than 1.");
//
//     this.player.one("seeked", _V_.proxy(this, function(){
//
//       _V_.log("seeked2", this.player.currentTime())
//
//       ok(this.player.currentTime() <= 1, "Time is less than 1.");
//       start();
//
//     }));
//
//     this.player.currentTime(0);
//
//   }));
//
//
//   this.player.play();
//
//   // this.player.one("timeupdate", _V_.proxy(this, function(){
//   //
//   //   this.player.currentTime(this.player.duration() / 2);
//   //
//   //   this.player.one("timeupdate", _V_.proxy(this, function(){
//   //     ok(this.player.currentTime() > 0, "Time is greater than 0.");
//   //
//   //     this.player.pause();
//   //     this.player.currentTime(0);
//   //
//   //     this.player.one("timeupdate", _V_.proxy(this, function(){
//   //
//   //       ok(this.player.currentTime() == 0, "Time is 0.");
//   //       start();
//   //
//   //     }));
//   //
//   //   }));
//   //
//   //
//   // }));
//
// });
//
// /* Events
// ================================================================================ */
// module("API Events", {
//   setup: playerSetup,
//   teardown: playerTeardown
// });
//
// var playEventList = []
//
// // Test all playback events
// test("Initial Events", 11, function() {
//   stop(); // Give 30 seconds to run then fail.
//
//   var events = [
//     // "loadstart" // Called during setup
//     "play",
//     "playing",
//
//     "durationchange",
//     "loadedmetadata",
//     "loadeddata",
//
//     "progress",
//     "timeupdate",
//
//     "canplay",
//     "canplaythrough",
//
//     "pause",
//     "ended"
//   ];
//
//   // Add an event listener for each event type.
//   for (var i=0, l=events.length; i<l; i++) {
//     var evt = events[i];
//
//     // Bind player and event name to function so event name value doesn't get overwritten.
//     this.player.one(evt, _V_.proxy({ player: this.player, evt: evt }, function(){
//       ok(true, this.evt);
//
//       // Once we reach canplaythrough, pause the video and wait for 'paused'.
//       if (this.evt == "canplaythrough") {
//         this.player.pause();
//
//       // After we've paused, go to the end of the video and wait for 'ended'.
//       } else if (this.evt == "pause") {
//         this.player.currentTime(this.player.duration() - 1);
//
//         // Flash has an issue calling play too quickly after currentTime. Hopefully we'll fix this.
//         setTimeout(this.player.proxy(function(){
//           this.play();
//         }), 250);
//
//       // When we reach ended, we're done. Continue with the test suite.
//       } else if (this.evt == "ended") {
//         start();
//       }
//     }));
//   }
//
//   this.player.play();
// });


build.sh
#!/bin/bash

# See README.md for instructions

pwd=`pwd`
video_js="$pwd/../video.js"
flex_sdk=/Applications/Flex

# use command-line parameters if the default locations
# aren't available
if [ ! -d "$video_js" ]
then
  video_js=$1
fi
if [ ! -d "$flex_sdk" ]
then
  flex_sdk=$2
fi

# if we needed command-line parameters and didn't get them,
# complain about it
if [ -z "$video_js" ] || [ -z "$flex_sdk" ]
then
  echo "Usage: setup.sh video_js_dir flex_sdk_dir"
  exit 1
fi

# complain about the incorrect directories passed in
if [ ! -d "$video_js" ]
then
  echo "video.js not found at $video_js"
  exit 1
fi
if [ ! -d "$flex_sdk" ]
then
  echo "Flex SDK not found at $flex_sdk"
  exit 1
fi

# and now we can get started, echoing out what we're doing

echo "Compiling bin-debug/VideoJS.swf..."
$flex_sdk/bin/mxmlc ./src/VideoJS.as -o ./bin-debug/VideoJS.swf -target-player=10.3 -library-path+=libs/HLSProvider.swc

echo "Copying VideoJS.swf to $video_js/src/swf/video-js.swf..."
cp bin-debug/VideoJS.swf "$video_js/src/swf/video-js.swf"

echo "Copying flash_demo.html and video-js.swf into $video_js/dist/video-js/..."
cp bin-debug/VideoJS.swf "$video_js/dist/video-js/video-js.swf"
cp flash_demo.html "$video_js/dist/video-js"

echo "Finished."

setup.sh
#!/bin/bash

# See README.md for instructions

video_js=$1
flex_sdk=$2

if [ -z "$video_js" ] || [ -z "$flex_sdk" ]
then
  echo "Usage: setup.sh video_js_dir flex_sdk_dir"
  exit 1
fi
if [ ! -d "$video_js" ]
then
  echo 'video.js not found at ' "$video_js"
  exit 1
fi
if [ ! -d "$flex_sdk" ]
then
  echo 'Flex SDK not found at ' "$flex_sdk"
  exit 1
fi

echo "Downloading the Flash 10.3 playerglobal.swc..." 
mkdir -p "$flex_sdk/frameworks/libs/player/10.3"
curl --silent -o "$flex_sdk/frameworks/libs/player/10.3/playerglobal.swc" "http://fpdownload.macromedia.com/get/flashplayer/updaters/10/playerglobal10_3.swc"

echo "Setting up Video.js and demo files in bin-debug..."
mkdir -p bin-debug
cp -r "$video_js/dist/video-js/" bin-debug
rm bin-debug/video-js.swf
cp -f demo.html bin-debug/

echo "Setting up Video.js and demo files in bin-release..."
mkdir -p bin-release
cp -rf bin-debug/ bin-release/
sed -i.bak 's/VideoJS.swf/video-js.swf/' bin-release/demo.html

echo "Finished."


Gruntfile.js
module.exports = function (grunt) {

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    connect: {
      dev: {
        port: 8000,
        base: '.'
      }
    },
    replace: {
      setupFlexConfig: {
        src: ['flex-config.xml'],
        dest: 'build/',
        replacements: [{
          from: '{FLEX_HOME}',
          to: '../node_modules/flex-sdk/lib/flex_sdk/frameworks/'
        },
          {
            from: '{TARGET_PLAYER}',
            to: '14.0'
          },
          {
            from: '{SWF_VERSION}',
            to: '25'
          }]
      }
    },
    mxmlc: {
      options: {
        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_16.html
        metadata: {
          // `-title "Adobe Flex Application"`
          title: 'VideoJS SWF',
          // `-description "http://www.adobe.com/flex"`
          description: 'http://www.videojs.com',
          // `-publisher "The Publisher"`
          publisher: 'Brightcove, Inc.',
          // `-creator "The Author"`
          creator: 'Brightcove, Inc.',
          // `-language=EN`
          // `-language+=klingon`
          language: 'EN',
          // `-localized-title "The Color" en-us -localized-title "The Colour" en-ca`
          localizedTitle: null,
          // `-localized-description "Standardized Color" en-us -localized-description "Standardised Colour" en-ca`
          localizedDescription: null,
          // `-contributor "Contributor #1" -contributor "Contributor #2"`
          contributor: null,
          // `-date "Mar 10, 2013"`
          date: null
        },

        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_18.html
        application: {
          // `-default-size 240 240`
          layoutSize: {
            width: 640,
            height: 360
          },
          // `-default-frame-rate=24`
          frameRate: 30,
          // `-default-background-color=0x869CA7`
          backgroundColor: 0x000000,
          // `-default-script-limits 1000 60`
          scriptLimits: {
            maxRecursionDepth: 1000,
            maxExecutionTime: 60
          }
        },

        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_19.html
        // `-library-path+=libraryPath1 -library-path+=libraryPath2`
        libraries: ['libs/*.*'],
        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_14.html
        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_17.html
        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_20.html
        // http://livedocs.adobe.com/flex/3/html/help.html?content=compilers_21.html
        compiler: {
          // `-accessible=false`
          'accessible': false,
          // `-actionscript-file-encoding=UTF-8`
          'actionscriptFileEncoding': null,
          // `-allow-source-path-overlap=false`
          'allowSourcePathOverlap': false,
          // `-as3=true`
          'as3': true,
          // `-benchmark=true`
          'benchmark': true,
          // `-context-root context-path`
          'contextRoot': null,
          // `-debug=false`
          'debug': false,
          // `-defaults-css-files filePath1 ...`
          'defaultsCssFiles': [],
          // `-defaults-css-url http://example.com/main.css`
          'defaultsCssUrl': null,
          // `-define=CONFIG::debugging,true -define=CONFIG::release,false`
          // `-define+=CONFIG::bool2,false -define+=CONFIG::and1,"CONFIG::bool2 && false"
          // `-define+=NAMES::Company,"'Adobe Systems'"`
          'defines': {},
          // `-es=true -as3=false`
          'es': false,
          // `-externs className1 ...`
          'externs': [],
          // `-external-library-path+=pathElement`
          'externalLibraries': [],
          'fonts': {
            // `-fonts.advanced-anti-aliasing=false`
            advancedAntiAliasing: false,
            // `-fonts.languages.language-range "Alpha and Plus" "U+0041-U+007F,U+002B"`
            // USAGE:
            // ```
            // languages: [{
            //   lang: 'Alpha and Plus',
            //   range: ['U+0041-U+007F', 'U+002B']
            // }]
            // ```
            languages: [],
            // `-fonts.local-fonts-snapsnot filePath`
            localFontsSnapshot: null,
            // `-fonts.managers flash.fonts.JREFontManager flash.fonts.BatikFontManager flash.fonts.AFEFontManager`
            // NOTE: FontManager preference is in REVERSE order (prefers LAST array item).
            //       For more info, see http://livedocs.adobe.com/flex/3/html/help.html?content=fonts_06.html
            managers: []
          },
          // `-incremental=false`
          'incremental': false
        }
      },
      videojs_swf: {
        files: {
          'dist/video-js.swf': ['src/VideoJS.as']
        }
      }
    },
    bumpup: {
      options: {
        updateProps: {
          pkg: 'package.json'
        }
      },
      file: 'package.json'
    },
    tagrelease: {
      file: 'package.json',
      commit:  true,
      message: 'Release %version%',
      prefix:  'v'
    },
    shell: {
      options: {
        failOnError: true
      },
      'git-diff-exit-code': { command: 'git diff --exit-code' },
      'git-diff-cached-exit-code': { command: 'git diff --cached --exit-code' },
      'git-add-dist-force': { command: 'git add dist --force' },
      'git-merge-stable': { command: 'git merge stable' },
      'git-merge-master': { command: 'git merge master' },
      'git-checkout-stable': { command: 'git checkout stable' },
      'git-checkout-master': { command: 'git checkout master' },
      'git-push-origin-stable': { command: 'git push origin stable' },
      'git-push-upstream-stable': { command: 'git push upstream stable' },
      'git-push-origin-master': { command: 'git push origin master' },
      'git-push-upstream-master': { command: 'git push upstream master' },
      'git-push-upstream-tags': { command: 'git push upstream --tags' }
    },
    prompt: {
      release: {
        options: {
          questions: [
            {
              config: 'release', // arbitray name or config for any other grunt task
              type: 'confirm', // list, checkbox, confirm, input, password
              message: 'You tested and merged the changes into stable?',
              default: false, // default value if nothing is entered
              // choices: 'Array|function(answers)',
              // validate: function(value){ console.log('hi', value); grunt.fatal('test'); return "error"; }, // return true if valid, error message if invalid
              // filter:  function(value), // modify the answer
              // when: function(answers) // only ask this question when this function returns true
            }
          ]
        }
      },
    }
  });

  grunt.loadNpmTasks('grunt-connect');
  grunt.loadNpmTasks('grunt-text-replace');
  grunt.loadNpmTasks('grunt-bumpup');
  grunt.loadNpmTasks('grunt-tagrelease');
  grunt.loadNpmTasks('grunt-npm');
  grunt.loadNpmTasks('grunt-shell');
  grunt.loadNpmTasks('grunt-prompt');
  grunt.loadNpmTasks('chg');

  grunt.registerTask('dist', ['replace', 'mxmlc']);
  grunt.registerTask('default', ['dist']);

  grunt.registerMultiTask('mxmlc', 'Compiling SWF', function () {
    // Merge task-specific and/or target-specific options with these defaults.
    var childProcess = require('child_process');
    var flexSdk = require('flex-sdk');
    var async = require('async');
    var pkg =  grunt.file.readJSON('package.json');

    var
      options = this.options,
      done = this.async(),
      maxConcurrency = 1,
      q,
      workerFn;

    workerFn = function(f, callback) {
      // Concat specified files.
      var srcList = f.src.filter(function(filepath) {
        // Warn on and remove invalid source files (if nonull was set).
        if (!grunt.file.exists(filepath)) {
          grunt.log.error('Source file "' + filepath + '" not found.');
          return false;
        }
        else {
          return true;
        }
      });

      var cmdLineOpts = [];

      if (f.dest) {
        cmdLineOpts.push('-output');
        cmdLineOpts.push(f.dest);
      }

      cmdLineOpts.push('-define=CONFIG::version, "' + pkg.version + '"');
      cmdLineOpts.push('-load-config+=build/flex-config.xml');
      cmdLineOpts.push('--');
      cmdLineOpts.push.apply(cmdLineOpts, srcList);

      grunt.verbose.writeln('package version: ' + pkg.version);
      grunt.verbose.writeln('mxmlc path: ' + flexSdk.bin.mxmlc);
      grunt.verbose.writeln('options: ' + JSON.stringify(cmdLineOpts));

      // Compile!
      childProcess.execFile(flexSdk.bin.mxmlc, cmdLineOpts, function(err, stdout, stderr) {
        if (!err) {
          grunt.log.writeln('File "' + f.dest + '" created.');
        }
        else {
          grunt.log.error(err.toString());
          grunt.verbose.writeln('stdout: ' + stdout);
          grunt.verbose.writeln('stderr: ' + stderr);

          if (options.force === true) {
            grunt.log.warn('Should have failed but will continue because this task had the `force` option set to `true`.');
          }
          else {
            grunt.fail.warn('FAILED');
          }

        }
        callback(err);
      });
    };

    q = async.queue(workerFn, maxConcurrency);
    q.drain = done;
    q.push(this.files);
  });

  /**
   * How releases work:
   *
   * Changes come from pullrequests to master or stable.
   * They are tested then pulled into their base branch.
   * A change log item is added to "Unreleased".
   * In a minor/major release, master is merged into stable
   *   (possibly by way of a release branch if testing more).
   *
   * Check out stable if not already checked out.
   * Run `grunt release:RELEASE_TYPE`
   *   RELEASE_TYPE = major, minor, or patch
   *   Does the following:
   *     Bump version
   *     Build dist
   *     Force add dist
   *     Rotate changelog
   *     Commit changes
   *     Tag release
   *
   *  Staging should be merged back into master.
   *  Push stable and master to origin.
   *  Run `npm publish`.
   */
  grunt.registerTask('release', 'Bump, build, and tag', function(type) {
    // major, minor, patch
    type = type ? type : 'patch';

    grunt.task.run([
      'prompt-release',                   // make sure user is ready
      'shell:git-diff-exit-code',         // ensure there's no unadded changes
      'shell:git-diff-cached-exit-code',  // ensure there's no added changes
      'shell:git-checkout-stable',        // must start on the stable branch
      'chg-release:'+type,                // add release to changelog
      'bumpup:'+type,                     // bump up the package version
      'dist',                             // build distribution
      'shell:git-add-dist-force',         // force add the distribution
      'tagrelease',                       // commit & tag the changes
      'shell:git-push-origin-stable',     // push changes to your fork
      'shell:git-push-upstream-stable',   // push changes to upstream
      'shell:git-push-upstream-tags',     // push version tag
      'npm-publish',                      // publish to npm
      'shell:git-checkout-master',        // switch to master branch
      'shell:git-merge-stable',           // merge stable into master
      'shell:git-push-origin-master',      // push changes to your fork
      'shell:git-push-upstream-master'    // push changes upstream
    ]);
  });

  // Can't bail out when a prompt-confirm returns false, so need this hack
  // https://github.com/dylang/grunt-prompt/issues/4
  grunt.registerTask('prompt-release', ['prompt:release', 'prompt-release-check']);
  grunt.registerTask('prompt-release-check', '', function(type) {
    if(!grunt.config('release')) {
      grunt.fatal('Confirmation failed.');
    }
  });
};
